alpaca/rest.go

```diff
@@ -742,6 +743,155 @@ func (c *Client) GetAsset(symbol string) (*Asset, error) {
 	return &asset, nil
 }
 
+const (
+	optionContractsRequestsMaxLimit = 10000
+)
+
+func setQueryLimit(q url.Values, totalLimit, pageLimit, received, maxLimit int) {
+	limit := 0 // use server side default if unset
+	if pageLimit != 0 {
+		limit = pageLimit
+	}
+	if totalLimit != 0 {
+		remaining := totalLimit - received
+		if remaining <= 0 { // this should never happen
+			return
+		}
+		if (limit == 0 || limit > remaining) && remaining <= maxLimit {
+			limit = remaining
+		}
+	}
+
+	if limit != 0 {
+		q.Set("limit", strconv.Itoa(limit))
+	}
+}
+
+type GetOptionContractsRequest struct {
+	UnderlyingSymbols     string
+	ShowDeliverable       bool
+	Status                OptionStatus
+	ExpirationDate        civil.Date
+	ExpirationDateGTE     civil.Date
+	ExpirationDateLTE     civil.Date
+	RootSymbol            string
+	Type                  OptionType
+	Style                 OptionStyle
+	StrikePriceGTE        decimal.Decimal
+	StrikePriceLTE        decimal.Decimal
+	PennyProgramIndicator bool
+	PageLimit             int
+	TotalLimit            int
+}
+
+// GetOptionContracts returns the list of Option Contracts.
+func (c *Client) GetOptionContracts(req GetOptionContractsRequest) ([]OptionContract, error) {
+	u, err := url.Parse(fmt.Sprintf("%s/%s/options/contracts", c.opts.BaseURL, apiVersion))
+	if err != nil {
+		return nil, err
+	}
+
+	q := u.Query()
+
+	if req.UnderlyingSymbols != "" {
+		q.Set("underlying_symbols", req.UnderlyingSymbols)
+	}
+
+	q.Set("show_deliverables", strconv.FormatBool(req.ShowDeliverable))
+
+	if req.Status != "" {
+		q.Set("status", string(req.Status))
+	}
+
+	if !req.ExpirationDate.IsZero() {
+		q.Set("expiration_date", req.ExpirationDate.String())
+	}
+
+	if !req.ExpirationDateGTE.IsZero() {
+		q.Set("expiration_date_gte", req.ExpirationDateGTE.String())
+	}
+
+	if !req.ExpirationDateLTE.IsZero() {
+		q.Set("expiration_date_lte", req.ExpirationDateLTE.String())
+	}
+
+	if req.RootSymbol != "" {
+		q.Set("root_symbol", req.RootSymbol)
+	}
+
+	if req.Type != "" {
+		q.Set("type", string(req.Type))
+	}
+
+	if req.Style != "" {
+		q.Set("style", string(req.Style))
+	}
+
+	if req.StrikePriceLTE.IsZero() {
```

Comment:
Your unit test actually caught this bug: should be negated. Same for `StrikePriceGTE`.

Reply:
Sigh, I'm sad I missed this. But that's one of the reasons we write unit tests.